dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxws.3.2/src/org/apache/cxf/jaxws/AbstractJAXWSMethodInvoker.java

```diff
@@ -0,0 +1,458 @@
+                }
+            }
+        }
+        return handlerScopedStuff;
+    }
+
+    protected void addHandlerProperties(WrappedMessageContext ctx,
+                                        Map<String, Object> handlerScopedStuff) {
+        
+        boolean isFinestEnabled = LOG.isLoggable(Level.FINEST);   // Liberty Change issue #26529
+        
+        for (Map.Entry<String, Object> key : handlerScopedStuff.entrySet()) {
+            ctx.put(key.getKey(), key.getValue(), Scope.HANDLER);
+            if (isFinestEnabled) {
+                LOG.finest("Added handler property to WrappedMessageContext with handler scope; key: " + key.getKey() + " ,value: " + key.getValue()); // Liberty Change issue #26529
+            }
+        }
+    }
+
+    private Message createResponseMessage(Exchange exchange) {
+        if (exchange == null) {
+            if (LOG.isLoggable(Level.FINEST)) {
+                LOG.finest("Exchange is null, can not create response message!"); // Liberty Change issue #26529
+            }
+            return null;
+        }
+        Message m = exchange.getOutMessage();
+        if (m == null && !exchange.isOneWay()) {
+            Endpoint ep = exchange.getEndpoint();
+            m = new MessageImpl();
+            m.setExchange(exchange);
+            m = ep.getBinding().createMessage(m);
+            exchange.setOutMessage(m);
+            if (LOG.isLoggable(Level.FINEST)) {
+                LOG.finest("Out message can not be found in exchange and it's not one way. New our message is created and added to exchange: " + m); // Liberty Change issue #26529
+            }
+        }
+        return m;
+    }
+
+    protected void updateWebServiceContext(Exchange exchange, MessageContext ctx) {
+        
+        boolean isFinestEnabled = LOG.isLoggable(Level.FINEST);   // Liberty Change issue #26529
+        // Guard against wrong type associated with header list.
+        // Need to copy header only if the message is going out.
+        if (ctx.containsKey(Header.HEADER_LIST)
+                && ctx.get(Header.HEADER_LIST) instanceof List<?>) {
+            List<?> list = (List<?>) ctx.get(Header.HEADER_LIST);
+            if (list != null && !list.isEmpty()) {
+                SoapMessage sm = (SoapMessage) createResponseMessage(exchange);
+                if (isFinestEnabled) {
+                    LOG.finest("MessageContext contains headers. Response SOAP message is been created. "); // Liberty Change issue #26529
+                }
+                if (sm != null) {
+                    Iterator<?> iter = list.iterator();
+                    while (iter.hasNext()) {
+                        sm.getHeaders().add((Header) iter.next());
+                        if (isFinestEnabled) {
+                            LOG.finest("Header found in MessageContext: " + sm.getHeaders().get(sm.getHeaders().size() - 1) + " is added to response SOAP message"); // Liberty Change issue #26529
```

Comment:
will this `sm.getHeaders().get(sm.getHeaders().size() - 1)` ever cause a NPE?

Reply:
Theoretically, It shouldn't cause NPE. But still changed to a safer design.